### PR TITLE
[REVIEW] Enable merge/set_index for cudf-based dataframes with cupy `values`

### PR DESCRIPTION
**Pull request by rjzamora**
This PR is a continuation of #5205, with the goal of allowing `rearrange_by_divisions`, `set_index`, `merge`, etc... to be used with `cudf`-based dataframes.  Due to recent progressin cudf, this requires two general changes:

1. cudf dataframe/series `.values` will soon return cupy (rather than numpy) arrays.  Since the pandas' `groupsort_indexer` function requires numpy, we need to check for a `values_host` attribute right before this function is called.  A cudf drop-in for `groupsort_indexer` is in progress, but the use of `values_host` is needed until then.

2. Since `hash_pandas_object` does not work for cudf dataframes,  we need to use `cudf.hash_columns` (when the attribute is available) within `partitioning_index`.  Note that this is required for `merge` to work with cudf-based dataframes.

cc @brandon-b-miller - These changes *should* allow dask_cudf tests to pass for [cudf#2655](https://github.com/rapidsai/cudf/pull/2655)

- [ ] Tests added / passed
- [ ] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by mrocklin** [comment]
Great, now we need to figure out where we can put code like this so that
we're guaranteed that it will be run.  @kkraus14 might have some thoughts

**Comment by rjzamora** [comment]
> 1. The failure manifests itself as things just hanging, is that correct? Is the Python process responsive? I usually test this by running a thread that watches the main thread

The python process does **not** seem to be responsive when there is a "hang".

> 2. Is it easy to pare down the problem to something that other folks can run easily?

Yes and no.  I am using the following code for testing/debugging:

```python
import dask.dataframe as dd
import dask
import cudf

ddf = dask.datasets.timeseries(dtypes={'name': str, 'id': int})
gddf = ddf.map_partitions(cudf.from_pandas)

gddf2 = dd.shuffle.set_index(gddf, "id", shuffle="tasks")
gddf2.compute()
```

The only "tricky" part is that the test requires my personal branch of `cudf` and `dask`:

- [cudf:rjzamora/merge-clean](https://github.com/rjzamora/cudf/tree/merge-clean) - Commit: e4abc7e66d5385e95ba4932eba237cf5ae6f82e4
- [dask:rjzamora/cudf-values-merge-cleanup](https://github.com/rjzamora/dask/tree/cudf-values-merge-cleanup) - Commit: ef9482b0b395b45dfc7deaec6f1b83994632ef2e

**Comment by mrocklin** [comment]
>> The failure manifests itself as things just hanging, is that correct? Is the Python process responsive? I usually test this by running a thread that watches the main thread

> The python process does not seem to be responsive when there is a "hang".

cc @quasiben @kkraus14 FYI .  This might be relevant for you all.

Does anyone have thoughts on how we can best figure out what is going on when things seize up?

**Comment by brandon-b-miller** [comment]
@rjzamora I think I have been able to reproduce the issue you're seeing using the same repo/commit setup you outlined above. One thing I noticed is that I can actually get your example to work if I massively downsample the rows to several hundred, and even then it takes 10-15 seconds.  Could this just be an extreme performance issue of some sort instead of a dead-stop hang?

To be specific I am adding basically a `ddf.sample(frac=somefrac)` before the `map_partitions` stage. I can get it to work for `somefrac=0.0001` which yields about 270 rows. Still probing it....

**Comment by rjzamora** [comment]
@brandon-b-miller Thanks for looking into this.  I don't *think* it is an extreme performance issue.  I can get the code to work fine (almost) when I write my own cupy-based version of `percentiles_summary` (replacing all numpy calls with cupy, etc).  I don't get any hanging when I run this version of the code once.  Unfortunately, things still seem to hang when I try using `%timeit`...

I didn't get to the bottom of that issue quite yet.  However, I do think this PR has dask is such a state that all/most of the cudf-related changes/fixes will need to be in cudf/dask_cudf.  The main remaining questions for this PR are:  (1) How should we name the new dispatch functions, and (2) where should they go?

**Comment by rjzamora** [comment]
cc @quasiben @kkraus14 @mrocklin @brandon-b-miller 

I have a gist with a not-so-simple reproducer [here](https://gist.github.com/rjzamora/098a2d89068d400c36b60260cf81928c).  The "hang" behavior can be reproduced with the master branches of cudf and dask.  Setting `use_cupy` to 0 or 3 should allow you to run through the cells without any problems.  Setting `use_cupy` to 1 or 2 should cause the code to **hang**. 

For now, I literally copied most of the `partition_quantiles` code from dask into the notebook.  My plan was to chip away at the logic to produce something smaller, but I had trouble with this.

It seems that the code hangs when a cupy array is used in `np.percentile`, and also when a `cupy` array is used in `cupy.percentile` (`use_cupy` == 1 or 2).  I tried doing a smaller experiment with `cupy.percentile` [here](https://gist.github.com/rjzamora/7a7e836f278cfdac606716d054d86951), but there was no hanging in those simpler cases.

**Comment by quasiben** [comment]
@rjzamora I am trying to look at this now -- currently getting an error with `use_cupy=1`.  I suspect this is because I am slightly behind branch-0.10 for cudf -- in the process of updating

**Comment by quasiben** [comment]
I am still not able to generate the hang but when I set the scheduler to processes:

```python
divisions, sizes, mins, maxes = base.compute(
        divisions, sizes, mins, maxes, optimize_graph=False, scheduler='processes'
    )
```

I get a CUDAERROR:

> CudaDriverError: CUDA initialized before forking

This was resolved in dask-cuda by initializing with numba:
https://github.com/rapidsai/dask-cuda/blob/d4d0d9f0ec764d3a87df712cb2c4aa965a088938/dask_cuda/initialize_context.py#L18-L27

Not sure how to get this function into the other processes dask spins up easily.  @mrocklin do you have thoughts ?

**Comment by quasiben** [comment]
@rjzamora I recently took a look at this PR: https://github.com/rapidsai/cudf/pull/2692 with my particular hang -- using is seemed to work though I don't understand why at the moment

**Comment by rjzamora** [comment]
@quasiben  Thanks for the info!  I will look into both the dask-cudf and cudf related leads

**Comment by rjzamora** [comment]
Update:  Thanks to a pointer from @brandon-b-miller I was able to dramatically reduce the size of the [`Hang_Reproducer` notebook](https://gist.github.com/rjzamora/098a2d89068d400c36b60260cf81928c).

It turns out that you can cause the hang by doing something as simple as `x = cupy.array([])` (without even using it) in `percentiles_summary`.

**Comment by brandon-b-miller** [comment]
Still looking into this on my side too. It appears that any call to CuPy that happens inside the actual tasks executed by the `queue.Queue` in `local.py` will cause the hang. I'm still running with the theory that there's either an issue where the threads are fighting over resources on the GPU in some way, or that some effect related to host/device synchronization isn't playing nicely with the asynchronous queue. 

I'm still in the process of putting print statements everywhere, will update this thread when I make progress.

**Comment by mrocklin** [comment]
I could believe that CuPy wasn't threadsafe.  It might be worth asking them directly on their issue tracker.

**Comment by brandon-b-miller** [comment]
The only thing that makes me really scratch my head is that I can set up a notebook where I make `queue.Queue` object with hundreds of threads, each of which make one hundred simple cupy calls as their 'task' - no hang. I'll verify and post an example.

**Comment by pentschev** [comment]
I've taken a look at the [notebook](https://gist.github.com/rjzamora/098a2d89068d400c36b60260cf81928c) and converted it into a python script to go a bit lower level for debugging. While it's very difficult to be precise on the causes and where the hang actually happens, gdb hints it's related to CUDA calls (some times I see memory allocation, memory freeing, memory copying, event handling, etc.) but not only in CuPy, threads also hang in RMM/cuDF calls.

I believe we only see the issues because we have various libraries running at the same time and there's some sort of incompatibility. For instance, there have been issues on multithreading reported in the past such as https://github.com/cupy/cupy/issues/1916 which is still open, but I can't even reproduce that. Apart from that, replacing the CuPy memory allocation with Numba memory allocation, there are no hangs.

So I'm wondering now, have you seen such hangs happening on a pipeline that doesn't include cuDF? If we could isolate that, it would certainly make debugging easier, but I'm fairly confident that the issue only happens because we're using both cuDF and CuPy together.

**Comment by rjzamora** [comment]
> I'm fairly confident that the issue only happens because we're using both cuDF and CuPy together.

This has certainly been my experience.  I have not experienced any hanging without cudf

**Comment by rjzamora** [comment]
I really appreciate that you took the time to look into this @pentschev !  Can you say what you did to test this: "Apart from that, replacing the CuPy memory allocation with Numba memory allocation, there are no hangs"

**Comment by mrocklin** [comment]
@kkraus14 you should probably also be aware of this.  It is maybe important for cudf generally, and possibly also has some impact on the UCX work.

**Comment by pentschev** [comment]
I've done a few tests, disabling both cuDF and CuPy memory pools, replacing `x = cupy.array([])` by several variations (`x = cupy.asarray([])`, `x = cupy.asarray(np.array([]))`, `x = np.random.random(1)`, perhaps a few others), and results are mostly consistent with hangs. Some of those may have caused hangs to be more sparse, but I can't tell if they were just coincidences or if they really cause less hangs. The only substitution that didn't hang a single time was plugging Numba instead of CuPy (`x = numba.cuda.to_device(np.array([]))`), which I would kinda expect since cuDF relies in part on that.

My best guess was to really disable all memory management, but that didn't seem to have much effect. I tried various combinations of that as well, playing with the following code and parameters:

```python
from librmm_cffi import librmm_config as rmm_cfg

rmm_cfg.use_pool_allocator = False  # default is False
rmm_cfg.initial_pool_size = 2<<30  # set to 2GiB. Default is 1/2 total GPU memory
rmm_cfg.use_managed_memory = False  # default is false
rmm_cfg.enable_logging = True      # default is False -- has perf overhead

import cudf
import cupy

cudf.rmm.finalize()
cudf.rmm.initialize()

cupy.cuda.set_allocator(None)
cupy.cuda.set_pinned_memory_allocator(None)
```

**Comment by rjzamora** [comment]
> My best guess was to really disable all memory management, but that didn't seem to have much effect.

Earlier today, @shwina suggested that there were currently problems with the `librmm_config` API -  I believe it prevents you from *enabling* rmm (rather than from disabling it), but I thought it was worth noting here.

**Comment by pentschev** [comment]
> Earlier today, @shwina suggested that there were currently problems with the librmm_config API - I believe it prevents you from enabling rmm (rather than from disabling it), but I thought it was worth noting here.

I don't know if there may be any issues with enabling rmm with cuDF, but it seems to be enabled by default (checked this on both 0.9 and 0.10 nightly):

```python
In [1]: import cudf                                                                                                                                                                                         

In [2]: cudf.rmm.is_initialized()                                                                                                                                                                           
Out[2]: True
```

**Comment by rjzamora** [comment]
@mrocklin -  Although we are currently using this PR to discuss the cupy+cudf "hang" problem, that "bug" should not be considered a good reason to hold up a merge here.

Merging this will break dask_cudf master, but only until the `hash_df`, `group_split`, and `group_split_2` dispatch functions are added for cudf types.  The cupy hang problem will not show up if the cudf definition of `group_split_2` avoids cupy (for now).

Since [cudf#2655](https://github.com/rapidsai/cudf/pull/2655) is already blocked by this PR, we can just add the dispatch functions there.

**Comment by pentschev** [comment]
> @mrocklin - Although we are currently using this PR to discuss the cupy+cudf "hang" problem, that "bug" should not be considered a good reason to hold up a merge here.

+1 to merging if that's not a blocker.

@rjzamora could you then start a new issue where we can continue to investigate/discuss the hang issue?

**Comment by mrocklin** [comment]
> Merging this will break dask_cudf master, but only until the hash_df, group_split, and group_split_2 dispatch functions are added for cudf types

Should we hold off until we have high confidence that those will go in?  Do we have that confidence now?

cc @kkraus14

**Comment by rjzamora** [comment]
> @rjzamora could you then start a new issue where we can continue to investigate/discuss the hang issue?

Yup - I will definitely open an issue when this is merged (I am thinking in cudf, since the problem is cudf specific)

> Should we hold off until we have high confidence that those will go in? Do we have that confidence now?

Right - Perhaps on Monday I can open a PR with the minimal changes needed in cudf.  Then we can merge here once there is more confidence all around :)

**Comment by rjzamora** [comment]
Note: [cudf#2809](https://github.com/rapidsai/cudf/pull/2809) includes the necessary cudf changes to address the changes in this PR

**Comment by rjzamora** [comment]
Discussion surrounding the cupy hang problem should now move to [cudf#2822](https://github.com/rapidsai/cudf/issues/2822)

**Comment by pentschev** [comment]
The hang is fixed by https://github.com/numba/numba/pull/4581, see https://github.com/rapidsai/cudf/issues/2822#issuecomment-532595103 for details.

**Comment by mrocklin** [comment]
It's nice to see that the numba change fixes both problems.

**Comment by rjzamora** [comment]
Thanks again @pentschev 

@mrocklin - Are there any remaining concerns/suggestions for the changes in this PR?

File: dask/dataframe/partitionquantiles.py [hunk 1]
```diff
@@ -385,7 +386,8 @@ def process_val_weights(vals_and_weights, npartitions, dtype_info):
     return rv
 
 
-def percentiles_summary(df, num_old, num_new, upsample, state):
+@methods.percentiles_summary.register(pd.Series)
+def percentiles_summary_pandas(df, num_old, num_new, upsample, state):
```
Comment (mrocklin) [inline dask/dataframe/partitionquantiles.py]: Is it possible to avoid dispatching this one? 

If we added a `tolist()` method to cudf, and allowed for a very small amount of host-device transfer for some random values, would we need to split this off?  Dispatch functions open up some code complexity.  I would prefer to use them sparingly if we can avoid them.  Obviously in some cases we can't avoid them, but this seems like a case where I'm not sure.  Thoughts?

File: dask/dataframe/methods.py [hunk 2]
```diff
@@ -480,7 +482,68 @@ def concat_pandas(dfs, axis=0, join="outer", uniform=False, filter_warning=True)
...
+    df2 = df.take(indexer)
+    locations = locations.cumsum()
+    parts = [df2.iloc[a:b] for a, b in zip(locations[:-1], locations[1:])]
+
+    return dict(zip(range(k), parts))
+
+
+# ---------------------------------
+# group_split_2
+# ---------------------------------
```
Comment (mrocklin) [inline dask/dataframe/methods.py]: We probably don't need separate comments like these for each function.  Maybe one larger header for "shuffling/merging/sorting" or something like that?

**Comment by mrocklin** [comment]
> @mrocklin - Are there any remaining concerns/suggestions for the changes in this PR?

Thanks for the work @rjzamora .  I gave this a closer look and raised a couple of small things.  In general most of this looks straightforward though.

cc @TomAugspurger , you might want to look at this.  We're making some of the shuffle code pandas/cudf dispatchable.  This probably increases maintenance burden a modest amount.

File: dask/dataframe/partitionquantiles.py (hunk 1 above)
Comment (rjzamora) [reply dask/dataframe/partitionquantiles.py]: Thanks for looking at this @mrocklin !

I added the `percentiles_summary` dispatch because the `data = df.values` line becomes problematic once the `values` property becomes a cupy array in cudf.

The original problem was that this leads to a "hang" during multi-threaded execution.  Therefore, I was planning to (temporarily) use `data = df.values_host` in the cudf definition. Assuming that the "hang" problem is fixed by numba#4581, the use of a cupy-typed `values` property will still cause a problem in `percentiles_summary`, because the required call to `np.percentile`  does **not** seem to handle cupy input.

Comment (rjzamora) [reply dask/dataframe/partitionquantiles.py]: If the number of new dispatch functions is a concern, it would be easiest for me to combine `group_split` and `group_split_2` (rather than removing `percentiles_summary`) -- Any thoughts on that?

**Comment by rjzamora** [comment]
@mrocklin - I was able to combine `group_split` and  `group_split_2` into a single function pretty easily.  It will be a challenge to avoid making `percentiles_summary` a dispatch function, but I can try to revise/rewrite the existing implementation to make it numpy/cupy agnostic (however, my personal opinion is that this is probably not worth the effort).

**Comment by mrocklin** [comment]
The group_split change looks nice!  Thanks!

I'm curious.  What stops percentiles_summary from being numpy/cupy agnostic?  I'm not trying to apply pressure in this direction, I'm just curious about what kinds of things still come up that make it hard to write backend agnostic code so that we can try to push on them upstream.

**Comment by rjzamora** [comment]
> What stops percentiles_summary from being numpy/cupy agnostic?

Last I checked,  using a cupy-based `df.values` led to an error within [this call to `np.percentile`](https://github.com/dask/dask/blob/6aab98cd4ef684cd79fb9390dc6ee7fbdce5adde/dask/array/percentile.py#L31).  There may have been another problem, but I'll need to test again (Note that my environment is not currently set up to run tests with cupy-`values`, but I am working on it).

**Comment by mrocklin** [comment]
What was the problem?  A quick spot check gave positive results:

```python
In [1]: import cupy, numpy

In [2]: x = cupy.arange(10000)

In [3]: numpy.percentile(x, [0, 50, 100])
Out[3]: array([   0. , 4999.5, 9999. ])

In [4]: type(_)
Out[4]: cupy.core.core.ndarray
```

My apologies also if you've already covered this above.  If you can point me to a traceback, I would be curious.

**Comment by rjzamora** [comment]
I just ran a similar test, and it worked for me as well :)

I will try `set_index` with the numba fix -  If I can get things working in that environment, I will be happy to remove the `percentiles_summary` dispatch

**Comment by rjzamora** [comment]
Okay - Using the numba fix does allow me to remove the `percentiles_summary` dispatch, but the preformance take a 20% hit.  I will investigate this and report back.

**Comment by rjzamora** [comment]
Using a clean environment with cupy-based `values` (and the numba fix in place) seems to work fine without the `percentiles_summary` dispatch -- Sorry about the confusion! (The problem I was seeing earlier was most likely caused by a workaround to the cupy hang that I was trying)

**Comment by mrocklin** [comment]
Great!  I'm glad to hear it.

File: dask/dataframe/shuffle.py [hunk 3]
```diff
@@ -562,22 +561,16 @@ def collect(p, part, meta, barrier_token):
...
         return {}, df
-    ind = df[col].values.astype(np.int64)
+    ind = df[col].astype(np.int64)
     n = ind.max() + 1
-    indexer, locations = groupsort_indexer(ind.view(np.int64), n)
-    df2 = df.take(indexer)
-    locations = locations.cumsum()
-    parts = [df2.iloc[a:b] for a, b in zip(locations[:-1], locations[1:])]
-    result2 = dict(zip(range(n), parts))
+    result2 = methods.group_split(df, ind.values.view(np.int64), n)
```
Comment (mrocklin) [inline dask/dataframe/shuffle.py]: Why these changes?  Why call astype before `.values` and why the call to `view`?

Comment (rjzamora) [reply dask/dataframe/shuffle.py]: When `values` is a cupy array, `ind = df[col].values.astype(np.int64)` does not result in a scalar value, but in a cupy array with one element. I could explicitly convert this cupy array to a scalar inside of `group_split_cupy`, but the change I currently have avoids the need for this.

I'm not really sure about the use of `view` - It was already in the code.

Comment (mrocklin) [reply dask/dataframe/shuffle.py]: Ah indeed, I see now that view is there regardless.  

Ah, I see about cupy.max, so you wanted cudf/pandas to call max, because presumably they'll return scalars while numpy/cupy diverge here.  That seems reasonable to me.  I was just curious.

Another option would be to explicitly call `int`, which is probably fine in both cases.

```python
In [1]: import cupy

In [2]: x = cupy.arange(10)

In [3]: x.max()
Out[3]: array(9)

In [4]: int(x.max())
Out[4]: 9
```

But I can't see much benefit to this over what you have here.  So I say that we just leave this as you have it here.

**Comment by mrocklin** [comment]
I'm +1 here.  Let's wait until tomorrow to make sure that @TomAugspurger has a chance to chime in if he wants to.

The CI failures seem unrelated.  It looks like maybe an upstream dependency changed.  Testing with an empty PR here: https://github.com/dask/dask/pull/5415

**Comment by rjzamora** [comment]
> I'm +1 here. Let's wait until tomorrow to make sure that @TomAugspurger has a chance to chime in if he wants to.

Sounds good to me - In the meantime it would be great to get some eyes on [cudf#2809](https://github.com/rapidsai/cudf/pull/2809) (the required cudf/dask_cudf component of this work).  It would be great to have the cudf dispatch definitions ready to merge when this PR goes in.

**Comment by TomAugspurger** [comment]
The changes here look good.

**Comment by mrocklin** [comment]
Merging this in.  Thanks for tracking this work @rjzamora !
